### PR TITLE
planner: consider more general cases when pushing lock down

### LINKED ISSUE #35524: Lock may not be pushed down to point get / batch point get in some cases
## Enhancement
image
The condition `len(p.Children()) > 1` can skip some cases even point get/batch point get exists.
For example, in the query statment `select * from t where id=1 for update union all select * from t where id = 2 for update`, the lock cannot be pushed down to point get as union all has two children.

If the lock is not pushed down to point get/batch point get, keys will not be locked in `lockKeyIfNeeded` (if the lock is push down to the point get/ batch point get, we should lock the keys here even though they are not exists). 

So, some race conditions can happen. 
For example, assume the table is empty and we execute 'begin pessimistic' in transaction 1. Then, during the execution of `select * from t where id=1 for update union all select * from t where id = 2 for update` and just after the point get plan get the latest ts, there can be a concurrent insert statement inserting some data including id 1 and id 2. We cannot get the data in transaction 1 nor block the concurrent insert statement.

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What problem does this PR solve?

Issue Number: close #35524

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: tests/realtikvtest/pessimistictest/pessimistic_test.go [hunk 1]
```diff
@@ -1332,6 +1332,85 @@ func TestPessimisticLockNonExistsKey(t *testing.T) {
...
+			tk.MustExec("commit")
+		} else {
+			select {
+			case <-ch:
+				break
+			case <-time.After(time.Second * 3):
+				panic("Timeout!")
+			}
+
+			tk.MustExec("commit")
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply tests/realtikvtest/pessimistictest/pessimistic_test.go]: Okay, it's done.

File: tests/realtikvtest/pessimistictest/pessimistic_test.go [hunk 2]
```diff
@@ -1332,6 +1332,85 @@ func TestPessimisticLockNonExistsKey(t *testing.T) {
 	tk1.MustExec("rollback")
 }
 
+func TestUnionAllPointGetLockBlockInsert(t *testing.T) {
+	store, _, clean := testkit.CreateMockStoreAndDomain(t)
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply tests/realtikvtest/pessimistictest/pessimistic_test.go]: Yes, we shoud use this.

File: planner/core/optimizer.go [hunk 3]
```diff
@@ -517,62 +517,118 @@ func physicalOptimize(logic LogicalPlan, planCounter *PlanCounterTp) (plan Physi
...
-			batchPointGet.Lock = lock
-			batchPointGet.LockWaitTime = waitTime
+
+	if len(physicalLockMap) != 0 {
+		for _, tuple := range tuples {
+			if tuple.physicalLock == nil {
+				continue
+			}
+
+			lock, waitTime := getLockWaitTime(sctx, tuple.physicalLock.Lock)
```
Comment (lcwangchao) [inline planner/core/optimizer.go:581]: I think the below plan is not right with the current implement. The table scan's lock was lost.
```
mysql> desc select * from (select * from t union all select * from t where id=1) t1 for update;
+----------------------------+---------+-----------+---------------+--------------------------------+
| id                         | estRows | task      | access object | operator info                  |
+----------------------------+---------+-----------+---------------+--------------------------------+
| Union_13                   | 3.00    | root      |               |                                |
| ├─TableReader_16           | 2.00    | root      |               | data:TableFullScan_15          |
| │ └─TableFullScan_15       | 2.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo |
| └─Point_Get_18             | 1.00    | root      | table:t       | handle:1, lock                 |
+----------------------------+---------+-----------+---------------+--------------------------------+
4 rows in set (0.00 sec)
```

File: planner/core/optimizer.go [hunk 4]
```diff
@@ -517,62 +517,118 @@ func physicalOptimize(logic LogicalPlan, planCounter *PlanCounterTp) (plan Physi
...
+			if !lock {
+				unionScan := physicalLockMap[tuple.physicalLock]
+				if unionScan != nil {
+					delete(unionScanSet, unionScan)
```
Comment (lcwangchao) [inline planner/core/optimizer.go:585]: Not very understand about this logic.

File: planner/core/optimizer.go [hunk 5]
```diff
@@ -517,62 +517,118 @@ func physicalOptimize(logic LogicalPlan, planCounter *PlanCounterTp) (plan Physi
...
+				}
+				if unionScan != nil {
+					unionScanSet[unionScan] = nil
+				}
+			case *PhysicalLock:
+				return x, nil, false
+			case *PhysicalUnionScan:
+				return nil, x, false
+			}
+			return nil, nil, false
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply planner/core/optimizer.go]: It's not needed now.

File: planner/core/optimizer.go [hunk 6]
```diff
@@ -517,62 +517,121 @@ func physicalOptimize(logic LogicalPlan, planCounter *PlanCounterTp) (plan Physi
...
+			} else {
+				tuple.batchPointGet.Lock = lock
+				tuple.batchPointGet.LockWaitTime = waitTime
+			}
 		}
 	}
+
 	return transformPhysicalPlan(p, func(p PhysicalPlan) PhysicalPlan {
-		if p == physLock {
+		if _, exist := physicalLockMap[p]; exist {
```
Comment (AilinKid) [inline planner/core/optimizer.go:607]: the old code is quite tricky, only take a single chain into consideration, once they found a pointGet or batchPointGet, that must be their leaf node(datasource). So just simply sink Lock OP to the inside of *PointGet.

Once remove the limitation for multi-child. For cases of Lock -> Union(pointGet, Selection+tableReader), we still need to keep the Lock OP above it to lock those keys selected from union's right executor.

File: tests/realtikvtest/pessimistictest/pessimistic_test.go [hunk 7]
```diff
@@ -1332,6 +1332,82 @@ func TestPessimisticLockNonExistsKey(t *testing.T) {
...
+			case <-ch:
+				panic("Should receive nothing")
+			case <-time.After(time.Second * 1):
+				break
+			}
+		} else {
+			select {
+			case <-ch:
+				break
+			case <-time.After(time.Second * 1):
```
Comment (TonsnakeLin) [inline tests/realtikvtest/pessimistictest/pessimistic_test.go:1398]: This line may lead to fail when the environment executing test cases runs slowly, I think this is one of the reasons that test cases run unstably on CI.
But I have no good idea.

File: planner/core/optimizer.go [hunk 8]
```diff
@@ -517,62 +517,121 @@ func physicalOptimize(logic LogicalPlan, planCounter *PlanCounterTp) (plan Physi
...
+	physicalLockMap := make(map[PhysicalPlan]PhysicalPlan)
+	unionScanSet := make(map[PhysicalPlan]interface{})
+
+	iteratePhysicalPlan(p, nil, nil,
+		func(p PhysicalPlan, physicalLock *PhysicalLock, unionScan *PhysicalUnionScan) (*PhysicalLock, *PhysicalUnionScan) {
+			// PhysicalUnionScan and PhysicalLock can only relate to one point get/batch point get.
+			// So, if the plan has multiple children, we should reset physicalLockMap and unionScanSet
+			if len(p.Children()) > 1 {
+				physicalLockMap = make(map[PhysicalPlan]PhysicalPlan)
+				unionScanSet = make(map[PhysicalPlan]interface{})
```
Comment (you06) [inline planner/core/optimizer.go:542]: So `iteratePhysicalPlan` performs a dfs and is there such a risk that we generate `physicalLockMap` and `unionScanSet` in one branch but reset them in another?

Comment (SpadeA-Tang) [reply planner/core/optimizer.go:542]: The intention of this is that we don't want one physical lock or union scan relates to mutiple point get/batch point get.

The current implementation has some problems still, I refactor it later.

**Pull request closed without merging**